clang/lib/Sema/SemaBounds.cpp

```diff
@@ -5623,6 +5712,84 @@ namespace {
       return Pair;
     }
 
+    // SynthesizeMembers modifies the set AbstractSets to include AbstractSets
+    // for member expressions whose target bounds use the value of the member
+    // expression M that is being modified via an assignment.
+    void SynthesizeMembers(Expr *E, MemberExpr *M, CheckedScopeSpecifier CSS,
+                           AbstractSetSetTy &AbstractSets) {
+      if (!E)
+        return;
+      Lexicographic Lex(S.Context, nullptr);
+      E = Lex.IgnoreValuePreservingOperations(S.Context, E->IgnoreParens());
+
+      // Recursively synthesize AbstractSets for certain subexpressions of E.
+      switch (E->getStmtClass()) {
+        case Expr::MemberExprClass:
+          break;
+        case Expr::ImplicitCastExprClass:
+        case Expr::CStyleCastExprClass: {
+          CastExpr *CE = cast<CastExpr>(E);
+          SynthesizeMembers(CE->getSubExpr(), M, CSS, AbstractSets);
+          return;
+        }
+        case Expr::UnaryOperatorClass: {
+          UnaryOperator *UO = cast<UnaryOperator>(E);
+          SynthesizeMembers(UO->getSubExpr(), M, CSS, AbstractSets);
+          return;
+        }
+        case Expr::ArraySubscriptExprClass: {
+          ArraySubscriptExpr *AE = cast<ArraySubscriptExpr>(E);
+          SynthesizeMembers(AE->getBase(), M, CSS, AbstractSets);
+          return;
+        }
+        case Expr::BinaryOperatorClass: {
+          BinaryOperator *BO = cast<BinaryOperator>(E);
+          Expr *LHS = BO->getLHS();
+          Expr *RHS = BO->getRHS();
+          if (LHS->getType()->isPointerType())
+            SynthesizeMembers(LHS, M, CSS, AbstractSets);
+          if (RHS->getType()->isPointerType())
+            SynthesizeMembers(RHS, M, CSS, AbstractSets);
+          return;
+        }
+        default:
+          return;
+      }
+
+      MemberExpr *ME = dyn_cast<MemberExpr>(E);
+      if (!ME)
+        return;
+
+      // Synthesize AbstractSets for a member expression `Base->Field`
+      // or `Base.Field`.
+      Expr *Base = ME->getBase();
+      FieldDecl *Field = dyn_cast<FieldDecl>(ME->getMemberDecl());
+      if (!Field)
+        return;
+
+      RecordDecl *Parent = Field->getParent();
+      if (!Parent)
+        return;
+
+      // Iterate over each sibling field F of Field to check whether the
+      // target bounds of `Base->F` or `Base.F` use the value of M.
+      for (auto It = Parent->field_begin(), E = Parent->field_end(); It != E; ++It) {
+        auto *F = *It;
+        if (!F->hasBoundsExpr())
+          continue;
+        MemberExpr *BaseF =
+          ExprCreatorUtil::CreateMemberExpr(S, Base, F, ME->isArrow());
+        BoundsExpr *Bounds = MemberExprTargetBounds(BaseF, CSS);
+        int Count = MemberOccurrenceCount(S, M, Bounds);
+        if (Count > 0) {
+          const AbstractSet *A = AbstractSetMgr.GetOrCreateAbstractSet(BaseF);
+          AbstractSets.insert(A);
+        }
```

Comment:
It will be great if we can add two profiling counters here: one for how many times lines 5780 - 5783 executed, and another one for how many times lines 5785 executed. These counters can be dumped after the compilation of the compilation unit.

Reply:
Are these counters something that should be added to a test file, e.g. synthesized-members.c (in which case it would be more straightforward to print them at the end of checking a function, similar to the existing dump for the synthesized abstract sets), or should they be more similar to the stats that are currently printed at the end of the compilation unit, e.g. in [S.getASTContext().PrintStats())](https://github.com/microsoft/checkedc-clang/blob/master/clang/lib/Parse/ParseAST.cpp#L184)?

Reply:
My thoughts: Similar to the stats that are currently printed at the end of the compilation unit e.g. in `S.getASTContext().PrintStats()`.